File: Framework/LanguageSupport/python/MMIPython/src/MMIPython/core/utils/thrift_client.py
## SPDX-License-Identifier: MIT
## The content of this file has been developed in the context of the MOSIM research project.
## Original author(s): Jannes Lehwald

# -*- coding: utf-8 -*-
"""

"""

from thrift.transport import TSocket
from thrift.transport import TTransport
from thrift.protocol import TCompactProtocol

class ThriftClient(object):
    """
    A wrapper for a service client connection
    
    Attributes
    ----------
    _address : str
        The address of the service
    _port : str
        The port of the service
    _transport : TTransport
        The thrift transport
    _access_type : type
        The type of the actual client
    _acces : Iface
        The actual access
    """
    
    def __init__(self, address, port, access_type):
        """
        Constructor which needs an address, a port and an access_type.
        
        Parameters
        ----------
        address : str
            The address of the service
        port : str
            The port of the service
        access_type : type
            The type of the actual client
        """
        
        assert(isinstance(address, str)), "The address is no string"
        assert(isinstance(port, int)), "The port is no int"
        assert(isinstance(access_type, type)), "The access_type is no type"
        
        self._address = address
        self._port = port
        self._transport = None
        self._access_type = access_type
        self._access = None
        
    def __enter__(self):
        """
        Used when entering pythons with statement.
        """
        self._transport = TSocket.TSocket(host=self._address, port=self._port)
        self._transport = TTransport.TBufferedTransport(self._transport)
        protocol = TCompactProtocol.TCompactProtocol(self._transport)
        
        self._access = self._access_type(protocol)
        
        self._transport.open()
        
        return self
    
    def __exit__(self, exc_type, exc_value, traceback):
        """
        Used when exiting pythons with statement.
        """
        self.dispose()
    
    def dispose(self):
        self._transport.close()
            
        

File: Framework/EngineSupport/Unreal/MOSIMPlugin/libs/includes/ThriftClient/ThriftClient.cpp
// SPDX-License-Identifier: MIT
// The content of this file has been developed in the context of the MOSIM research project.
// Original author(s): Andreas Kaiser, Niclas Delfs, Stephan Adam

#ifndef _THRIFT_CLIENT_CPP_
#define _THRIFT_CLIENT_CPP_

#include "ThriftClient.h"
#include <thrift/protocol/TCompactProtocol.h>
#include <thrift/transport/TSocket.h>
#include <thrift/transport/TTransportUtils.h>
#include "Utils/Logger.h"
//#include "boost/exception/diagnostic_information.hpp"

using namespace std;
using namespace apache::thrift::protocol;
using namespace MMIStandard;

template<class T>
 ThriftClient<T>::ThriftClient(std::string const & address, int port, bool autoOpen) :port(port), address(address)
{
	 shared_ptr<TTransport> socket(new TSocket(this->address, this->port));
	 this->transport = shared_ptr<TTransport>(new TBufferedTransport(socket));
	 shared_ptr<TProtocol> protocol(new TCompactProtocol{ transport });
	 this->access = shared_ptr<T>(new T{ protocol });

	 if (autoOpen)
		 this->Start();
}

template<class T>
ThriftClient<T>::~ThriftClient()
{
	this->transport->close();
}

template<class T>
void ThriftClient<T>::Start()
{
	try {
		if(!this->transport->isOpen())
			this->transport->open();
	}
	catch (exception e)
	{
		//Logger::printLog(L_ERROR, boost::current_exception_diagnostic_information());
		Logger::printLog(L_ERROR, e.what());	// TODO: define error message
	}
	
}
#endif